File: py/bud_sdk/features.py
# Bud SDK feature factory

from bud_sdk.feature.base_feature import BudBaseFeature
from bud_sdk.feature.debug_feature import BudDebugFeature
from bud_sdk.feature.idempotency_feature import BudIdempotencyFeature
from bud_sdk.feature.metrics_feature import BudMetricsFeature
from bud_sdk.feature.paging_feature import BudPagingFeature
from bud_sdk.feature.ratelimit_feature import BudRatelimitFeature
from bud_sdk.feature.retry_feature import BudRetryFeature
from bud_sdk.feature.test_feature import BudTestFeature
from bud_sdk.feature.timeout_feature import BudTimeoutFeature


_FEATURES = {
    "base": lambda: BudBaseFeature(),
    "debug": lambda: BudDebugFeature(),
    "idempotency": lambda: BudIdempotencyFeature(),
    "metrics": lambda: BudMetricsFeature(),
    "paging": lambda: BudPagingFeature(),
    "ratelimit": lambda: BudRatelimitFeature(),
    "retry": lambda: BudRetryFeature(),
    "test": lambda: BudTestFeature(),
    "timeout": lambda: BudTimeoutFeature(),
}


def _make_feature(name):
    factory = _FEATURES.get(name)
    if factory is not None:
        return factory()
    return _FEATURES["base"]()


# True when this SDK was generated with the named feature class - the
# constructor's tolerance for extend-carried features reads this (an
# active name with no generated class must not become a BaseFeature
# stray when an extend instance carries it).
def _has_feature(name):
    return name in _FEATURES

File: php/features.php
<?php
declare(strict_types=1);

// Bud SDK feature factory

require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/feature/DebugFeature.php';
require_once __DIR__ . '/feature/IdempotencyFeature.php';
require_once __DIR__ . '/feature/MetricsFeature.php';
require_once __DIR__ . '/feature/PagingFeature.php';
require_once __DIR__ . '/feature/RatelimitFeature.php';
require_once __DIR__ . '/feature/RetryFeature.php';
require_once __DIR__ . '/feature/TestFeature.php';
require_once __DIR__ . '/feature/TimeoutFeature.php';


class BudFeatures
{
    public static function make_feature(string $name)
    {
        switch ($name) {
            case "base":
                return new BudBaseFeature();
            case "debug":
                return new BudDebugFeature();
            case "idempotency":
                return new BudIdempotencyFeature();
            case "metrics":
                return new BudMetricsFeature();
            case "paging":
                return new BudPagingFeature();
            case "ratelimit":
                return new BudRatelimitFeature();
            case "retry":
                return new BudRetryFeature();
            case "test":
                return new BudTestFeature();
            case "timeout":
                return new BudTimeoutFeature();
            default:
                return new BudBaseFeature();
        }
    }

    /**
     * Does a generated feature class back this name? False for a name only
     * an options extend instance can supply (the station adopt path) - the
     * constructor uses this to skip make_feature for such names instead of
     * adding a stray BaseFeature.
     */
    public static function has_feature(string $name): bool
    {
        switch ($name) {
            case "base":
            case "debug":
            case "idempotency":
            case "metrics":
            case "paging":
            case "ratelimit":
            case "retry":
            case "test":
            case "timeout":
                return true;
            default:
                return false;
        }
    }
}
